src/cryptonote_core/blockchain.cpp

```diff
@@ -1900,6 +1900,55 @@ bool Blockchain::get_blocks(uint64_t start_offset, size_t count, std::vector<std
   return true;
 }
 //------------------------------------------------------------------
+bool Blockchain::is_request_sane(const NOTIFY_REQUEST_GET_OBJECTS::request& arg, std::string &error_message) const
+{
+  LOG_PRINT_L3("Blockchain::" << __func__);
+
+  if (arg.blocks.empty())
+  {
+    error_message = "NOTIFY_REQUEST_GET_OBJECTS is empty";
+    return false;
+  }
+  if (arg.blocks.size() > std::max<size_t>(BLOCKS_SYNCHRONIZING_DEFAULT_COUNT_PRE_V4, BLOCKS_SYNCHRONIZING_DEFAULT_COUNT))
+  {
+    error_message = "NOTIFY_REQUEST_GET_OBJECTS is too large (" + std::to_string(arg.blocks.size()) + ")";
+    return false;
+  }
+
+  CRITICAL_REGION_LOCAL(m_blockchain_lock);
+  const uint64_t blockchain_height = m_db->height();
+
+  uint64_t height = m_db->get_block_height(arg.blocks.front());
+  if (height + arg.blocks.size() - 1 >= blockchain_height)
+  {
+    error_message = "NOTIFY_REQUEST_GET_OBJECTS requests objects past the end of our blockchain (assuming contiguous)";
+    return false;
+  }
+
+  // based off borromean proofs usage
+  static const uint64_t v4_height = m_hardfork->get_earliest_ideal_height_for_version(4);
+  static const uint64_t v8_height = m_hardfork->get_earliest_ideal_height_for_version(8);
+  const size_t max_count = height < v4_height ? BLOCKS_SYNCHRONIZING_DEFAULT_COUNT_PRE_V4 : BLOCKS_SYNCHRONIZING_DEFAULT_COUNT;
+  if (arg.blocks.size() > max_count)
+  {
+    error_message = "NOTIFY_REQUEST_GET_OBJECTS requests more objects then we allow (" + std::to_string(arg.blocks.size()) + " requested, " + std::to_string(max_count) + " allowed)";
+    return false;
+  }
+
+  for (size_t i = 1; i < arg.blocks.size(); ++i)
+  {
+    const uint64_t height_i = m_db->get_block_height(arg.blocks[i]);
+    if (height_i != height + 1)
+    {
+      error_message = "NOTIFY_REQUEST_GET_OBJECTS requests non contiguous blocks (height " + std::to_string(height_i) + " following height " + std::to_string(height) + ")";
```

Comment:
Think we shouldn't forbid non contiguous get blocks requests.

It is just the way Monero reference implementation (`monreod`) currently works and I doubt we should enforce this on the protocol level.

Could you clarify the problem/weakness here?

Reply:
It prevents the (admittedly fairly coarse) check against pre-v4 and post v4 max blocks from being useful.

Reply:
This doesn't help to mitigates a DoS anyhow. Am i missing something?

Reply:
If blocks aren't contiguous, you can't easily clamp to 20 (instead of 100) for post rct.

Reply:
and if you're going to say "count bytes", it's not a trivial patch.

Reply:
It just breaks P2P protocol for no good reason.

Bad actor will just DoS you with contiguous blocks request.

Reply:
I was not intended to break anything. Can you be more specific ?

Reply:
`NOTIFY_REQUEST_GET_OBJECTS` request contains a list of block hashes to receive. It puts no restriction on the heights.
So it allows me to request block #0 and block #100 in a single request.

You are going to forbid this and force p2p client to always request contiguous blocks, i.e. block #0, block #1, block #2

What is the point of having a list of block hashes in the request then? Just a the `first block hash` and `blocks count` would be enough.


Summing this up. The patch introduces some *problems* to legit P2P clients and changes almost nothing from the DoS perspective, an attacker will just select 100 contiguous blocks span and spam your node with the requests.

Reply:
The proper way would be just to change the server code. We should just limit the blocks in the reply. I.e. no matter how many blocks a client requested, we always serve only first `N` blocks from the request, `N` - the limit we want to put there (20/100/500/etc.).

Reply:
monerod does not do that (unless you find a case where it does).
The point of having a list of block hashes is that it is a lot more work to avoid doing that. WhatI don't like the argument of saying no improvement because a better improvement can be made regardless of the amount of work involved (in general, it can apply in some cases). There wlil be a new call where you give a starting hash and a number, but I object to having to do everything at once or nothing.
The patch does not introduce problems to legit P2P clients, unless you find a case where it does.
The patch does mitigate DoS potential somewhat (I do not claim it removes the DoS) because it does not allow more than 20 blocks for the borromean part of the chain where blocks are large, and no more than 100 (which is the max monerod requests unless you change it manually) otherwise, rather than 500.

Reply:
It is all about P2P protocol, not about *referene monero node implementation* (monerod).

There *might be* other monero node implementations. And this change will break the P2P protocol for them. Even if you still want to do this, you should bump the P2P protocol version.

But i already mentioned the correct way this should be handled, see https://github.com/monero-project/monero/pull/6554#discussion_r426601668

Reply:
I really do not care. If anyone made custom code, they can just change the number of blocks they have.
You can't just send fewer blocks, the client will complain and drop. This will require something else, which I will *later*.